src/util/circular_buffer.c

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright Redis Ltd. 2016 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+
+#include "redismodule.h"
+#include "circular_buffer.h"
+#include "src/rmalloc.h"
+#include "src/module.h"
+
+#include <stdatomic.h>
+
+// Circular buffer structure.
+// The buffer is of fixed size.
+// Items are removed by order of insertion, similar to a queue.
+struct _CircularBuffer {
+	char *read;                   // read data from here
+	_Atomic uint64_t write;       // write offset into data
+	size_t item_size;             // item size in bytes
+	_Atomic uint64_t item_count;  // current number of items in buffer
+	uint64_t item_cap;            // max number of items held by buffer
+	char *end_marker;             // marks the end of the buffer
+	char data[];                  // data
+};
+
+// Creates a new circular buffer, with `cap` items of size `item_size`
+CircularBuffer CircularBuffer_New
+(
+	size_t item_size,  // size of item in bytes
+	uint cap           // max number of items in buffer
+) {
```

Comment:
e.g:
```suggestion
CircularBuffer CircularBuffer_New(size_t item_size,  uint cap) {
```

Reply:
Ah this.. I actually like it, but it is indeed a different format from what we currently have.
I think we should use this, but I can align until we apply the formatter.